charms/kfp-api/src/charm.py

```diff
@@ -486,7 +486,15 @@ def _get_relational_db_data(self) -> dict:
         self._get_db_relation("relational-db")
 
         # retrieve database data from library
-        relation_data = self.database.fetch_relation_data()
+        try:
+            # if called in response to a '*-relation-broken' event, this will raise an exception
+            relation_data = self.database.fetch_relation_data()
+        except KeyError:
```

Comment:
To understand this change, could you please explain how and under which circumstances the KeyError will be raised when calling `self.database.fetch_relation_data()`?
I don't really see anything [here](https://github.com/canonical/kfp-operators/blob/0580c89e0f4143c6e781fcc4ec6ae384675c08ce/charms/kfp-api/lib/charms/data_platform_libs/v0/data_interfaces.py#L511) that could potentially raise this exception.

Reply:
So, they say in the docstring that:
> Function cannot be used in `*-relation-broken` events and will raise an exception.

In their code, they do the following:
```python
relation.data[relation.app]
```

Apparently, on `relation-broken` `relation.app` can be `None`, which will raise a `KeyError` if used to index the dict with.

This can be seen in the trace provided in the linked issue: https://github.com/canonical/kfp-operators/issues/222

Reply:
Now, note, I haven't been able to reproduce this with Juju 2.9.44. I'm not sure if that's still the expected behaviour (i.e. `relation.app` becoming `None` when the relation breaks), so I implemented this based on the trace I got.

Reply:
(from an offline conversation) I think it is okay to handle this exception here as the library doesn't for us (we should probably raise an issue in the library so they handle this better?), but IMHO to fix the issue we should be paying more attention to this:

* The error suggests that there was something wrong when handling the relational-db relation broken event`2023-06-20T10:02:41.420Z [container-agent] 2023-06-20 10:02:41 DEBUG juju-log relational-db:18: Emitting Juju event relational_db_relation_broken.`
* That event's handler is [this](https://github.com/canonical/kfp-operators/blob/main/charms/kfp-api/src/charm.py#L666) 👿 
* The above will put the unit to maintenance and then call `_on_event()` which will do ALL the things that are done regularly (get interfaces, generate config, upload files to container, etc.)
* Because this is a relation broken event, some of the data that should be in the relation is not there, and thus some of these calls in `on_event()` will fail, specially [`config_json = self._generate_config(interfaces)`](https://github.com/canonical/kfp-operators/blob/main/charms/kfp-api/src/charm.py#L676) because that one will call [`get_db_data()`](https://github.com/canonical/kfp-operators/blob/main/charms/kfp-api/src/charm.py#L517), which also calls [this](https://github.com/canonical/kfp-operators/blob/main/charms/kfp-api/src/charm.py#L531), which will eventually call [`fetch_relation_data`](https://github.com/canonical/kfp-operators/blob/main/charms/kfp-api/src/charm.py#L489), the method that cannot be used in relation broken event by design.

I think for instance [this](https://github.com/canonical/kfp-operators/blob/main/charms/kfp-api/src/charm.py#L489) comment is more or less valid and I'd probably look at [this](https://github.com/canonical/kfp-operators/blob/main/charms/kfp-api/src/charm.py#L643) handler and try to make it better to handle the case of relation broken.

Reply:
The description above is accurate. I'm not sure about the last 2 links, don't think they point to the intended thing.

Currently, we use a single handler to tackle both the `relation-departed` and `relation-broken` events. Now, I believe that `relation-departed` is handled properly. We go through the same actions we do for most events, and generating the configuration will clear the fields that correspond to the departed relation. So that is actually fine.

What do we really expect the charm to do on a `relation-broken` event, though? Given that it always comes after the departure has been handled, the corresponding relation databag fields have already been cleared. This is all about the config anyway. These relations do not affect the K8s resources applied whatsoever. So what our handler will do with the proposed fix is realise that the relation is no more available and set the charm to `Blocked`, with the `"Please add required database relation: eg. relational-db"` message.

Do you have a different behaviour in mind? Is there something else that we should be doing on `relation-broken` events?

Reply:
Sorry about the links, I have updated them in my previous comment.

> What do we really expect the charm to do on a relation-broken event, though?

I guess this is the important question here. This topic is something we have been discussing in the team for a while. I remember we had this conversation when @ca-scribner was designing the [dashboard library](https://github.com/canonical/kubeflow-dashboard-operator/blob/main/lib/charms/kubeflow_dashboard/v0/kubeflow_dashboard_links.py). Though it is different than this problem, I think it provides valuable information, and you can definitely take a look at it for inspiration.
To be honest, I would have to think on what should be done in the relation broken event handler to provide a good answer.

> These relations do not affect the K8s resources applied whatsoever. So what our handler will do with the proposed fix is realise that the relation is no more available and set the charm to Blocked, with the "Please add required database relation: eg. relational-db" message.
Do you have a different behaviour in mind? Is there something else that we should be doing on relation-broken events?

It is true the relations (parting or being added) do not affect the K8s resources applied (that is why I said the comment was more or less accurate).

Now, deleting the application with `juju remove kfp-api` is triggering a relation broken event (?), which will then call `on_event()`, calling this will cause the same actions we are doing on install or on config change to be run. This is not desired because on remove you want to remove things, not render + apply + generate config (one of the methods that is raising) + fetch relation data (one of the methods that is raising), etc.

While your proposed solution will fix the issue in the case where you remove the DB application (`mysql-k8s`) and there is a sudden relation broken, it will put the unit to BlockedStatus on remove, something that wouldn't make sense because you want to remove, not relate it to another application.

Reply:
Oh I get what you're saying.

I'm hesitant about refactoring the remove handler as part of this effort, because then we would also have to update the `joined`/`changed` handlers which don't need to apply K8s resources either, and probably others as well, since now all of them call `_on_event`.

However, we could try setting the `db_data` to something empty after catching the error instead of setting the unit in Blocked status.

I think that what you're describing though is mostly a matter of semantics. Yes, it doesn't make sense for `kfp-api` to go into `Blocked` state asking for relations when we try to remove it. However, if we deal with the raised `KeyError`, this would never actually cause an issue, being just a transient state (and most likely imperceptible on the user side). I mean, Juju will update the relation first, firing the relevant events (i.e. `departed` -> `broken`), but then it will actually proceed with the application removal, which means that the `_on_remove` handler will run at the end, deleting any K8s resources created.

Reply:
So yes, I'm all for restructuring our handlers so that they're more fine-grained and perform the exact tasks that should concern them. However, IMO this is a much greater effort design and implementation-wise, given that I'm not sure any of our charms actually go for that approach.

Taking a look at the dashboard library.

Reply:
@phoevos has provided with unit and integration tests that can help confirm the exception will be handled and won't interfere in the removal process. resolving this conversation.